quic/s2n-quic-platform/src/io/tokio.rs

```diff
@@ -545,70 +517,48 @@ impl<E: Endpoint<PathHandle = PathHandle>> Instance<E> {
-                let subscriber = endpoint.subscriber();
-                let mut publisher = event::EndpointPublisherSubscriber::new(
-                    event::builder::EndpointMeta {
-                        endpoint_type: E::ENDPOINT_TYPE,
-                        timestamp: clock.get_time(),
-                    },
-                    None,
-                    subscriber,
-                );
-
-                publisher.on_platform_event_loop_wakeup(event::builder::PlatformEventLoopWakeup {
-                    timeout_expired: timeout_result,
-                    rx_ready: rx_result.is_some(),
-                    tx_ready: tx_result.is_some(),
-                });
-
-                if let Some(guard) = tx_result {
-                    if let Ok(result) = guard?.try_io(|socket| tx.tx(socket, &mut publisher)) {
-                        result?;
-                    }
-                }
+            let (rx_result, tx_result, timeout_expired, application_wakeup) =
+                if let Ok(res) = Select::new(rx_task, tx_task, &mut wakeups, &mut timer).await {
+                    res
+                } else {
+                    // The endpoint has shut down
+                    return Ok(());
+                };
 
-                // if the timer expired ensure it is reset to the default timeout
-                if timeout_result {
-                    reset_clock = true;
+            let subscriber = endpoint.subscriber();
+            let mut publisher = event::EndpointPublisherSubscriber::new(
+                event::builder::EndpointMeta {
+                    endpoint_type: E::ENDPOINT_TYPE,
+                    timestamp: clock.get_time(),
+                },
+                None,
+                subscriber,
+            );
+
+            publisher.on_platform_event_loop_wakeup(event::builder::PlatformEventLoopWakeup {
+                timeout_expired,
+                rx_ready: rx_result.is_some(),
+                tx_ready: tx_result.is_some(),
+                application_wakeup,
+            });
+
+            if let Some(guard) = tx_result {
+                if let Ok(result) = guard?.try_io(|socket| tx.tx(socket, &mut publisher)) {
+                    result?;
                 }
+            }
 
-                if let Some(guard) = rx_result {
-                    if let Ok(result) = guard?.try_io(|socket| rx.rx(socket, &mut publisher)) {
-                        result?;
-                    }
-                    endpoint.receive(&mut rx.rx_queue(), &clock);
+            if let Some(guard) = rx_result {
+                if let Ok(result) = guard?.try_io(|socket| rx.rx(socket, &mut publisher)) {
+                    result?;
                 }
-            } else {
-                // The endpoint has shut down
-                return Ok(());
+                endpoint.receive(&mut rx.rx_queue(), &clock);
             }
 
             endpoint.transmit(&mut tx.tx_queue(), &clock);
```

Comment:
just curious why we rx before tx and if we tried the other way around? also if one order is better than other for client vs server.

Reply:
We technically do both "concurrently". If we have capacity in the Rx buffer, we poll the socket. If we have a packet to send, we poll the tx socket. But this transmit call needs to be called each iteration, since transmission can be caused by all of the other events (app wakeup, timeout, receiving a packet). Whereas the Rx call only happens when a packet comes in.